### DISCUSSION TITLE
Send DICOM via HTTP Sender with custom body

### DISCUSSION THREAD

**Discussion by alliedarmour** [nextgenhealthcare/connect#4572] [Q&A]
I'm relatively new to mirth and we're developing kind of a data warehouse for DICOM files and metadata. I've set up my API and now I want to send local stored DICOM images via the HTTP Sender.

Sending the file as form data seems to work, but is it possible to change the request body according to my specifications in the API?

For example I expect the following format: `{ "dicom": { "dicom_file": **Encoded DICOM FIle here** } }`

I tried it with setting the content in the channel destination like this `{ "dicom": { "dicom_file": "${message.rawData}" } }` but it seems that the structure was ignored, at on my API Endpoint I just received the multipart file.

Thanks for any help!

**Comment by pacmano1** [comment]
I think you have to share your http receiver code to answer that question since that defines how post to it.

The fact it is DICOM is mostly irrelevant here.   Although I might expect DICOM images sizes to be large enough to require chunking, but I'll defer to others here on that part.

**Reply by alliedarmour** [reply]
> I think you have to share your http receiver code to answer that question since that defines how post to it.
> 
> The fact it is DICOM is mostly irrelevant here. Although I might expect DICOM images sizes to be large enough to require chunking, but I'll defer to others here on that part.

I don't have an HTTP Receiver, I'm building an API with Ruby on Rails and I expect the data to be in this certain format. The problem is that the key which points to the multipart file is the temporary filename (which changes from file to file). Is there any way to define custom keys directly in the Content section when using POST multipart/form-data or do you have to build the object through a transformer with javascript?

Thanks for your answer!

**Comment by pacmano1** [comment]
For posting multipart form data you have to do something like this, obviously I have some vars set elsewhere in the below earlier in the channel and I am using an attachment handler for the pdf itself.
```javascript
$c('boundary','--------'+UUIDGenerator.getUUID())

var pdfContent = getAttachments(true).get(0).getContent();
var entity = org.apache.http.entity.mime.MultipartEntityBuilder
    .create()
    .addTextBody('patient', $('fn_mrn'))
    .setBoundary($('boundary'))
    .addTextBody('description', 'Consolidated Evidence of Care')
    .setBoundary($('boundary'))
    .addTextBody('metatags', '["Some  Documents"]')
    .setBoundary($('boundary'))
    .addTextBody('doctor', $('provider_external_id'))
    .setBoundary($('boundary'))
    .addTextBody('date', $c('billing_period_end_datetime'))
    .setBoundary($('boundary'))
    .addTextBody('document', $('fn'))
    .setBoundary($('boundary'))
    .addBinaryBody('document', pdfContent, org.apache.http.entity.ContentType.create('application/octet-stream'), $('fn'))       
    .build();

var bos = new java.io.ByteArrayOutputStream();
entity.writeTo(bos);
msg = FileUtil.encode(bos.toByteArray());

**Reply by pacmano1** [reply]
LOL, couldn't figure out code highlighting...

**Reply by pacmano1** [reply]
Credit to @tonygermano and @kirbykn2 (initial solution over at the forums).

**Reply by pacmano1** [reply]
If you are just assembling JSON for a non-multipart post, in general I use transformers for that.

e.g. an outbound message template:

```json
{
    "dicom": {
        "dicom_file": ""
    }
}
```
and in a transformer step for that destination:
```javascript
tmp['dicom']['dicom_file'] = // the encoded file
```
of course you can just do that in the destination content on the destination tab itself.

**Reply by alliedarmour** [reply]
Okay I tried a bit and came up with a solution which doesn't really seem to work the way I want. I built a "custom" multipart HTTP request with setting multipart to NO in the destination HTTP sender. Then I defined a boundary variable in the transformer and put it in the destination map. My problem now is: When I try to send the Content in the content field of the destination connector, it is just empty. This is my channel config:

![channel](https://user-images.githubusercontent.com/20596363/120248747-bdbbd600-c278-11eb-889a-5b2d7ff2fd7b.PNG)

But the params are always empty, any idea?
